src/tree/updater_quantile_hist.cc

```diff
@@ -152,26 +203,51 @@ void QuantileHistMaker::Builder::BuildHistogramsLossGuide(
   int starting_index = std::numeric_limits<int>::max();
   int sync_count = 0;
 
-  AddHistRows(&starting_index, &sync_count);
+  AddHistRows(&starting_index, &sync_count, p_tree);
   BuildLocalHistograms(gmat, gmatb, p_tree, gpair_h);
   SyncHistograms(starting_index, sync_count, p_tree);
 }
 
 
-void QuantileHistMaker::Builder::AddHistRows(int *starting_index, int *sync_count) {
+void QuantileHistMaker::Builder::AddHistRows(int *starting_index, int *sync_count,
```

Comment:
Unfortunately this function became massive and needed to be refactored.
I think we can add firstly left and secondary right nodes - in simpler manner.

For example, you don't need to create tmp vector and sort it. It's enough to just have a loop which handle all required things.

Reply:
unfortunately it should be sorted, without sorting I get wrong results(different evaluation error)